src/send-linux.c

```diff
@@ -0,0 +1,115 @@
+{
+	// Get the actual socket
+	int sock = s.sock;
+	// get source interface index
+	struct ifreq if_idx;
+	memset(&if_idx, 0, sizeof(struct ifreq));
+	if (strlen(zconf.iface) >= IFNAMSIZ) {
+		log_error("send", "device interface name (%s) too long\n",
+			  zconf.iface);
+		return EXIT_FAILURE;
+	}
+	strncpy(if_idx.ifr_name, zconf.iface, IFNAMSIZ - 1);
+	if (ioctl(sock, SIOCGIFINDEX, &if_idx) < 0) {
+		log_error("send", "%s", "SIOCGIFINDEX");
+		return EXIT_FAILURE;
+	}
+	int ifindex = if_idx.ifr_ifindex;
+
+	// destination address for the socket
+	memset((void *)&sockaddr, 0, sizeof(struct sockaddr_ll));
+	sockaddr.sll_ifindex = ifindex;
+	sockaddr.sll_halen = ETH_ALEN;
+	if (zconf.send_ip_pkts) {
+		sockaddr.sll_protocol = htons(ETHERTYPE_IP);
+	}
+	memcpy(sockaddr.sll_addr, zconf.gw_mac, ETH_ALEN);
+	return EXIT_SUCCESS;
+}
+
+int send_batch(sock_t sock, batch_t* batch, int retries) {
+	struct mmsghdr msgvec [batch->capacity]; // Array of multiple msg header structures
+	struct msghdr msgs[batch->capacity];
+	struct iovec iovs[batch->capacity];
+
+	for (int i = 0; i < batch->len; ++i) {
+		struct iovec *iov = &iovs[i];
+	    	iov->iov_base = ((void *)batch->packets) + (i * MAX_PACKET_SIZE);
+	       	iov->iov_len = batch->lens[i];
+		struct msghdr *msg = &msgs[i];
+		memset(msg, 0, sizeof(struct msghdr));
+		// based on https://github.com/torvalds/linux/blob/master/net/socket.c#L2180
+		msg->msg_name = (struct sockaddr *)&sockaddr;
+		msg->msg_namelen = sizeof(struct sockaddr_ll);
+		msg->msg_iov = iov;
+		msg->msg_iovlen = 1;
+		msgvec[i].msg_hdr = *msg;
+		msgvec[i].msg_len = batch->lens[i];
+	}
+	// set up per-retry variables, so we can only re-submit what didn't send successfully
+	struct mmsghdr* current_msg_vec = msgvec;
+	int total_packets_sent = 0;
+	int num_of_packets_in_batch = batch->len;
+	for (int i = 0; i < retries; i++) {
+		// according to manpages
+		// On success, sendmmsg() returns the number of messages sent from msgvec; if this is less than vlen, the
+		//       caller can retry with a further sendmmsg() call to send the remaining messages.
+		// On error, -1 is returned, and errno is set to indicate the error.
+		int rv = sendmmsg(sock.sock, current_msg_vec, num_of_packets_in_batch, 0);
+		if (rv < 0) {
+			// retry if sending all packets failed
+			log_error("batch send", "error in sendmmsg: %s", strerror(errno));
+			continue;
+		}
+		// if rv is positive, it gives the number of packets successfully sent
+		total_packets_sent += rv;
+		if (rv == num_of_packets_in_batch){
+			// all packets in batch were sent successfully
+			break;
+		}
+		// batch send was only partially successful, we'll retry if we have retries available
+		log_warn("batch send", "only successfully sent %d packets out of a batch of %d packets", total_packets_sent, batch->len);
+		// remove successfully sent packets from batch for retry
+		current_msg_vec = &msgvec[total_packets_sent];
+		num_of_packets_in_batch = batch->len - total_packets_sent;
```

Comment:
This looks like the it assumes the failed packets are always the back of `msgvec`...is that accurate? Could an arbitrary subset fail?

If it's always the back of the vector that failed, then I would recommend adding a comment to clarify.

Reply:
Yep, that was _my_ interpretation of the [man pages](https://man7.org/linux/man-pages/man2/sendmmsg.2.html),
```
If an error occurs after at least one message has been sent, the
       call succeeds, and returns the number of messages sent.  The
       error code is lost.  The caller can retry the transmission,
       starting at the first failed message, but there is no guarantee
       that, if an error is returned, it will be the same as the one
       that was lost on the previous call.
```

So  I don't believe an arbitrary subset could fail. I'll add a clarifying comment to call this out!